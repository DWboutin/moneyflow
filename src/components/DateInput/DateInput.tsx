import React, { useEffect, useState } from "react";
import ReactDatePicker from "react-datepicker";
import styled from "styled-components";
import { ThemeContainer } from "../../styles/styles";

interface ContainerProps {}

export const Container = styled.div<ContainerProps>`
  display: flex;
  flex-direction: column;
  flex: 1;

  label {
    margin-bottom: 10px;
  }

  .react-datepicker {
    border: 0;
    border-radius: 0 20px 20px 20px;
    box-shadow: ${({ theme }: ThemeContainer) => theme.card.boxShadow};

    &__triangle {
      &::before {
        border-bottom: 0px !important;
      }
      &::after {
        border-bottom-color: ${({ theme }: ThemeContainer) =>
          theme.datepicker.header.bg} !important;
      }
    }

    &__header {
      background-color: ${({ theme }: ThemeContainer) =>
        theme.datepicker.header.bg};
      border-radius: 0 20px 0 0 !important;

      * {
        color: ${({ theme }: ThemeContainer) => theme.datepicker.header.text};
      }
    }

    &__month-container {
      border: 0;
    }

    &__day--selected,
    &__day--keyboard-selected {
      background-color: ${({ theme }: ThemeContainer) =>
        theme.datepicker.selectedDay};
    }

    &__navigation-icon::before {
      border-color: ${({ theme }: ThemeContainer) =>
        theme.datepicker.header.text};
    }
  }
`;

const Input = styled(ReactDatePicker)`
  height: 40px;
  width: 100%;
  padding: 10px 20px 9px 20px;
  border-radius: 0px 20px 20px 20px;
  color: ${({ theme }: ThemeContainer) => theme.input.text};
  border: 1px solid ${({ theme }: ThemeContainer) => theme.input.normal.border};
  background-color: ${({ theme }: ThemeContainer) => theme.input.normal.bg};

  &::placeholder {
    color: ${({ theme }: ThemeContainer) => theme.input.placeholder};
  }

  &:focus-visible {
    border: 1px solid ${({ theme }: ThemeContainer) => theme.input.focus.border};
    outline: none;
  }
`;

const Error = styled.div`
  padding-top: 5px;
  color: ${({ theme }: ThemeContainer) => theme.input.error};
`;

export interface Props {
  id: string;
  name: string;
  label: string;
  placeholder: string;
  handleOnChange: (date: Date | null) => void;
  error?: string | boolean;
}

const TextInput: React.VoidFunctionComponent<Props> = ({
  id,
  name,
  label,
  placeholder,
  handleOnChange,
  error,
}) => {
  const [date, setDate] = useState<Date | null>(new Date());

  useEffect(() => {
    handleOnChange(date);
  }, [date]);

  return (
    <Container>
      <label htmlFor={id}>{label}</label>
      <Input
        id={id}
        name={name}
        selected={date}
        maxDate={new Date()}
        onChange={(date: Date) => setDate(date)}
        placeholderText={placeholder}
      />
      {error && <Error>{error}</Error>}
    </Container>
  );
};

export default TextInput;
