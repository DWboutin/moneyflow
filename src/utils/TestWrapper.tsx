import React, { ReactNode } from "react";
import { Provider } from "react-redux";
import { ToastContainer } from "react-toastify";
import { ThemeProvider } from "styled-components";

import store from "../store/store";
import GlobalStyle from "../styles/globalStyle";

import { theme } from "../styles/styles";

interface Props {
  children: ReactNode;
}

const TestWrapper: React.FunctionComponent<Props> = ({ children }) => (
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <ToastContainer />
    </ThemeProvider>
  </Provider>
);

export default TestWrapper;
