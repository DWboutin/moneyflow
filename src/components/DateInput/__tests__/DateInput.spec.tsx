import React from "react";
import { fireEvent, render, RenderResult, act } from "@testing-library/react";

import DateInput, { Props } from "../DateInput";
import TestWrapper from "../../../utils/TestWrapper";

describe("<DateInput />", () => {
  let renderResult: RenderResult;

  const renderDateInput = ({
    id,
    name,
    label,
    placeholder,
    handleOnChange,
    error,
  }: Props) => {
    renderResult = render(
      <TestWrapper>
        <DateInput
          id={id}
          name={name}
          label={label}
          placeholder={placeholder}
          handleOnChange={handleOnChange}
          error={error}
        />
      </TestWrapper>
    );
  };

  const mockOnChange = jest.fn();

  describe("default rendering", () => {
    beforeEach(() => {
      renderDateInput({
        id: "id",
        name: "name",
        label: "Input label",
        placeholder: "Date placeholder",
        handleOnChange: mockOnChange,
      });
    });

    it("should match snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });
  });

  describe("with error", () => {
    beforeEach(() => {
      renderDateInput({
        id: "id",
        name: "name",
        label: "Input label",
        placeholder: "Date placeholder",
        handleOnChange: mockOnChange,
        error: "Input error",
      });
    });

    it("should match snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });
  });

  describe("when clicking the input", () => {
    beforeEach(async () => {
      renderDateInput({
        id: "id",
        name: "name",
        label: "Input label",
        placeholder: "Date placeholder",
        handleOnChange: mockOnChange,
      });
      fireEvent(
        renderResult.container.querySelector("input") as HTMLInputElement,
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
        })
      );

      await act(async () => await null); // from github issue https://github.com/floating-ui/floating-ui/issues/1520
    });

    it("should match snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });

    it("should the date picker be rendered", async () => {
      expect(
        renderResult.container.querySelector(".react-datepicker__tab-loop")
      ).not.toBe(null);
    });

    it("should execute handleOnChange when changing date", () => {
      fireEvent(
        renderResult.container.querySelector(
          ".react-datepicker__tab-loop .react-datepicker__week .react-datepicker__day:first-child"
        ) as HTMLInputElement,
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
        })
      );

      expect(mockOnChange).toHaveBeenCalled();
    });
  });
});
