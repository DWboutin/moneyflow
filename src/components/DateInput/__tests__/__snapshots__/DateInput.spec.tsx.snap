// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateInput /> default rendering should match snapshot 1`] = `
<div>
  <div
    class="sc-bdvvtL jIUZAq"
  >
    <label
      for="id"
    >
      Input label
    </label>
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container"
      >
        <input
          class="sc-gsDKAQ cJKViw"
          id="id"
          name="name"
          placeholder="Date placeholder"
          type="text"
          value="03/11/2022"
        />
      </div>
    </div>
  </div>
  <div
    class="Toastify"
  />
</div>
`;

exports[`<DateInput /> when clicking the input should match snapshot 1`] = `
<div>
  <div
    class="sc-bdvvtL jIUZAq"
  >
    <label
      for="id"
    >
      Input label
    </label>
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container"
      >
        <input
          class="sc-gsDKAQ cJKViw react-datepicker-ignore-onclickoutside"
          id="id"
          name="name"
          placeholder="Date placeholder"
          type="text"
          value="03/11/2022"
        />
      </div>
    </div>
    <div
      class="react-datepicker__tab-loop"
    >
      <div
        class="react-datepicker__tab-loop__start"
        tabindex="0"
      />
      <div
        class="react-datepicker-popper"
        data-placement="bottom-start"
        style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
      >
        <div>
          <div
            class="react-datepicker"
          >
            <div
              class="react-datepicker__triangle"
              style="position: absolute; left: 0px; transform: translate(0px, 0px);"
            />
            <button
              aria-label="Previous Month"
              class="react-datepicker__navigation react-datepicker__navigation--previous"
              type="button"
            >
              <span
                class="react-datepicker__navigation-icon react-datepicker__navigation-icon--previous"
              >
                Previous Month
              </span>
            </button>
            <div
              class="react-datepicker__month-container"
            >
              <div
                class="react-datepicker__header "
              >
                <div
                  class="react-datepicker__current-month"
                >
                  March 2022
                </div>
                <div
                  class="react-datepicker__header__dropdown react-datepicker__header__dropdown--scroll"
                />
                <div
                  class="react-datepicker__day-names"
                >
                  <div
                    class="react-datepicker__day-name"
                  >
                    Su
                  </div>
                  <div
                    class="react-datepicker__day-name"
                  >
                    Mo
                  </div>
                  <div
                    class="react-datepicker__day-name"
                  >
                    Tu
                  </div>
                  <div
                    class="react-datepicker__day-name"
                  >
                    We
                  </div>
                  <div
                    class="react-datepicker__day-name"
                  >
                    Th
                  </div>
                  <div
                    class="react-datepicker__day-name"
                  >
                    Fr
                  </div>
                  <div
                    class="react-datepicker__day-name"
                  >
                    Sa
                  </div>
                </div>
              </div>
              <div
                aria-label="month  2022-03"
                class="react-datepicker__month"
                role="listbox"
              >
                <div
                  class="react-datepicker__week"
                >
                  <div
                    aria-disabled="false"
                    aria-label="Choose Sunday, February 27th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--027 react-datepicker__day--weekend react-datepicker__day--outside-month"
                    role="option"
                    tabindex="-1"
                  >
                    27
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Monday, February 28th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--028 react-datepicker__day--outside-month"
                    role="option"
                    tabindex="-1"
                  >
                    28
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Tuesday, March 1st, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--001"
                    role="option"
                    tabindex="-1"
                  >
                    1
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Wednesday, March 2nd, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--002"
                    role="option"
                    tabindex="-1"
                  >
                    2
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Thursday, March 3rd, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--003"
                    role="option"
                    tabindex="-1"
                  >
                    3
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Friday, March 4th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--004"
                    role="option"
                    tabindex="-1"
                  >
                    4
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Saturday, March 5th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--005 react-datepicker__day--weekend"
                    role="option"
                    tabindex="-1"
                  >
                    5
                  </div>
                </div>
                <div
                  class="react-datepicker__week"
                >
                  <div
                    aria-disabled="false"
                    aria-label="Choose Sunday, March 6th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--006 react-datepicker__day--weekend"
                    role="option"
                    tabindex="-1"
                  >
                    6
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Monday, March 7th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--007"
                    role="option"
                    tabindex="-1"
                  >
                    7
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Tuesday, March 8th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--008"
                    role="option"
                    tabindex="-1"
                  >
                    8
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Wednesday, March 9th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--009"
                    role="option"
                    tabindex="-1"
                  >
                    9
                  </div>
                  <div
                    aria-disabled="false"
                    aria-label="Choose Thursday, March 10th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--010"
                    role="option"
                    tabindex="-1"
                  >
                    10
                  </div>
                  <div
                    aria-current="date"
                    aria-disabled="false"
                    aria-label="Choose Friday, March 11th, 2022"
                    aria-selected="true"
                    class="react-datepicker__day react-datepicker__day--011 react-datepicker__day--selected react-datepicker__day--today"
                    role="option"
                    tabindex="0"
                  >
                    11
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Saturday, March 12th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--012 react-datepicker__day--disabled react-datepicker__day--weekend"
                    role="option"
                    tabindex="-1"
                  >
                    12
                  </div>
                </div>
                <div
                  class="react-datepicker__week"
                >
                  <div
                    aria-disabled="true"
                    aria-label="Not available Sunday, March 13th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--013 react-datepicker__day--disabled react-datepicker__day--weekend"
                    role="option"
                    tabindex="-1"
                  >
                    13
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Monday, March 14th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--014 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    14
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Tuesday, March 15th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--015 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    15
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Wednesday, March 16th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--016 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    16
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Thursday, March 17th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--017 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    17
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Friday, March 18th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--018 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    18
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Saturday, March 19th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--019 react-datepicker__day--disabled react-datepicker__day--weekend"
                    role="option"
                    tabindex="-1"
                  >
                    19
                  </div>
                </div>
                <div
                  class="react-datepicker__week"
                >
                  <div
                    aria-disabled="true"
                    aria-label="Not available Sunday, March 20th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--020 react-datepicker__day--disabled react-datepicker__day--weekend"
                    role="option"
                    tabindex="-1"
                  >
                    20
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Monday, March 21st, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--021 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    21
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Tuesday, March 22nd, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--022 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    22
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Wednesday, March 23rd, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--023 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    23
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Thursday, March 24th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--024 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    24
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Friday, March 25th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--025 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    25
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Saturday, March 26th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--026 react-datepicker__day--disabled react-datepicker__day--weekend"
                    role="option"
                    tabindex="-1"
                  >
                    26
                  </div>
                </div>
                <div
                  class="react-datepicker__week"
                >
                  <div
                    aria-disabled="true"
                    aria-label="Not available Sunday, March 27th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--027 react-datepicker__day--disabled react-datepicker__day--weekend"
                    role="option"
                    tabindex="-1"
                  >
                    27
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Monday, March 28th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--028 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    28
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Tuesday, March 29th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--029 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    29
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Wednesday, March 30th, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--030 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    30
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Thursday, March 31st, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--031 react-datepicker__day--disabled"
                    role="option"
                    tabindex="-1"
                  >
                    31
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Friday, April 1st, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--001 react-datepicker__day--disabled react-datepicker__day--outside-month"
                    role="option"
                    tabindex="-1"
                  >
                    1
                  </div>
                  <div
                    aria-disabled="true"
                    aria-label="Not available Saturday, April 2nd, 2022"
                    aria-selected="false"
                    class="react-datepicker__day react-datepicker__day--002 react-datepicker__day--disabled react-datepicker__day--weekend react-datepicker__day--outside-month"
                    role="option"
                    tabindex="-1"
                  >
                    2
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="react-datepicker__tab-loop__end"
        tabindex="0"
      />
    </div>
  </div>
  <div
    class="Toastify"
  />
</div>
`;

exports[`<DateInput /> with error should match snapshot 1`] = `
<div>
  <div
    class="sc-bdvvtL jIUZAq"
  >
    <label
      for="id"
    >
      Input label
    </label>
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container"
      >
        <input
          class="sc-gsDKAQ cJKViw"
          id="id"
          name="name"
          placeholder="Date placeholder"
          type="text"
          value="03/11/2022"
        />
      </div>
    </div>
    <div
      class="sc-dkPtRN AKYuc"
    >
      Input error
    </div>
  </div>
  <div
    class="Toastify"
  />
</div>
`;
